### PR TITLE
Prevent deletion of repositories that are used by snapshot backed indices

### PR DESCRIPTION
**Pull request by tlrx**
This commit adds a simple check that prevents to delete a repository that is required by an existing searchable snapshot index within the same cluster.

Note to reviewer: I labeled this pull request as an `>enhancement` but one can argue that repository deletion not being forbidden is a bug. I'm tempted to backport this change to 7.x as it is.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java [hunk 1]
```diff
@@ -625,45 +630,70 @@ private static void validate(final String repositoryName) {
...
+                + " searchable snapshots indices that use the repository: "
+                + Strings.collectionToCommaDelimitedString(indices)
+                + (count > indices.size() ? ",..." : "")
+            );
+        }
+    }
+
+    private static boolean indexSettingsMatchRepositoryMetadata(Settings indexSettings, RepositoryMetadata repositoryMetadata) {
+        if ("snapshot".equals(INDEX_STORE_TYPE_SETTING.get(indexSettings))) {
+            return Objects.equals(repositoryMetadata.uuid(), indexSettings.get("index.store.snapshot.repository_uuid"))
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java:683]: Nit: I'd prefer if we made these string literals into constants in this class and then referred to them from `SearchableSnapshots` when declaring the settings.

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java [hunk 2]
```diff
@@ -625,45 +630,70 @@ private static void validate(final String repositoryName) {
...
+                || Objects.equals(repositoryMetadata.name(), indexSettings.get("index.store.snapshot.repository_name"));
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java:684]: If `indexSettings.get("index.store.snapshot.repository_uuid")` is nonempty then we shouldn't care about the repo name I think. E.g. in Cloud it should be ok to remove/adjust the `found-snapshots` repository if all your searchable snapshots are from a different cluster's `found-snapshots` repo which you're calling something different.

**Review by DaveCTurner — COMMENTED** [review]
Looks good, I left a couple of small comments.

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java (hunk 2 above)
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java]: It makes sense, thanks for spotting this. I pushed 700dffb738f3fee73c9349d7de94428c10965823

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java (hunk 1 above)
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java]: Sure, I pushed 928b3983a770834ede527bce5b178f1d2a3fed5d

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by tlrx** [comment]
Thanks @DaveCTurner ! I think it can go as it is to `7.14` unless you think this is too breaking?

**Comment by DaveCTurner** [comment]
Ah sorry I missed that this was a question. Yes, I think this is reasonable to call a bugfix, certainly ok by me to backport to 7.x.

**Comment by tlrx** [comment]
@DaveCTurner merci :)

**Pull request merged**